File: packages/python/elizaos/basic_capabilities/providers/character.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_provider_spec
from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_provider_spec("CHARACTER")


def _resolve_name(text: str, name: str) -> str:
    """Replace ``{{name}}`` placeholders with the character's name.

    Supports character template files where the name is injected at render
    time so changing the character's name doesn't require rewriting every
    field.
    """
    return text.replace("{{name}}", name)


def _resolve_name_list(items: list[str], name: str) -> list[str]:
    """Resolve ``{{name}}`` in every element of a string list."""
    return [_resolve_name(s, name) for s in items]


async def get_character_context(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    character = runtime.character
    agent_name: str = character.name

    sections: list[str] = []

    sections.append(f"# Agent: {agent_name}")

    if character.bio:
        if isinstance(character.bio, str):
            bio_text = _resolve_name(character.bio, agent_name)
        else:
            bio_text = "\n".join(_resolve_name_list(list(character.bio), agent_name))
        sections.append(f"\n## Bio\n{bio_text}")

    if character.adjectives:
        adjectives = (
            character.adjectives
            if isinstance(character.adjectives, list)
            else [character.adjectives]
        )
        resolved_adjectives = _resolve_name_list(adjectives, agent_name)
        sections.append(f"\n## Personality Traits\n{', '.join(resolved_adjectives)}")

    # lore is optional and may not exist on all Character instances
    lore = getattr(character, "lore", None)
    if lore:
        if isinstance(lore, str):
            lore_text = _resolve_name(lore, agent_name)
        else:
            lore_text = "\n".join(_resolve_name_list(list(lore), agent_name))
        sections.append(f"\n## Background\n{lore_text}")

    if character.topics:
        topics = character.topics if isinstance(character.topics, list) else [character.topics]
        resolved_topics = _resolve_name_list(topics, agent_name)
        sections.append(f"\n## Knowledge Areas\n{', '.join(resolved_topics)}")

    if character.style:
        style_sections: list[str] = []
        if character.style.all:
            all_style = (
                character.style.all
                if isinstance(character.style.all, list)
                else [character.style.all]
            )
            resolved_all = _resolve_name_list(all_style, agent_name)
            style_sections.append(f"General: {', '.join(resolved_all)}")
        if character.style.chat:
            chat_style = (
                character.style.chat
                if isinstance(character.style.chat, list)
                else [character.style.chat]
            )
            resolved_chat = _resolve_name_list(chat_style, agent_name)
            style_sections.append(f"Chat: {', '.join(resolved_chat)}")
        if character.style.post:
            post_style = (
                character.style.post
                if isinstance(character.style.post, list)
                else [character.style.post]
            )
            resolved_post = _resolve_name_list(post_style, agent_name)
            style_sections.append(f"Posts: {', '.join(resolved_post)}")
        if style_sections:
            sections.append("\n## Communication Style\n" + "\n".join(style_sections))

    context_text = "\n".join(sections)

    # Note: Protobuf ProviderResult.data is a Struct which has limited type support.
    # The text already contains all the information needed for the agent context.
    return ProviderResult(text=context_text)


character_provider = Provider(
    name=_spec["name"],
    description=_spec["description"],
    get=get_character_context,
    dynamic=_spec.get("dynamic", False),
)

File: packages/typescript/src/basic-capabilities/providers/character.ts
import {
  buildDeterministicSeed,
  deterministicHex,
  deterministicPickOne,
  deterministicSample,
} from "../../deterministic";
import { requireProviderSpec } from "../../generated/spec-helpers.ts";
import type {
  IAgentRuntime,
  Memory,
  Provider,
  State,
} from "../../types/index.ts";
import { ChannelType } from "../../types/index.ts";
import { addHeader } from "../../utils.ts";

// Get text content from centralized specs
const spec = requireProviderSpec("CHARACTER");

/**
 * Replace `{{name}}` placeholders in a string with the character's name.
 * Supports character template files where the name is injected at render time
 * so changing the character's name doesn't require rewriting every field.
 */
function resolveNamePlaceholder(text: string, name: string): string {
  return text.replaceAll("{{name}}", name);
}

/** Resolve `{{name}}` in every element of a string array. */
function resolveNameInArray(items: string[], name: string): string[] {
  return items.map((s) => resolveNamePlaceholder(s, name));
}

/**
 * Character provider object.
 * @typedef {Object} Provider
 * @property {string} name - The name of the provider ("CHARACTER").
 * @property {string} description - Description of the character information.
 * @property {Function} get - Async function to get character information.
 */
/**
 * Provides character information.
 * @param {IAgentRuntime} runtime - The agent runtime.
 * @param {Memory} message - The message memory.
 * @param {State} state - The state of the character.
 * @returns {Object} Object containing values, data, and text sections.
 */
export const characterProvider: Provider = {
  name: spec.name,
  description: spec.description,
  get: async (runtime: IAgentRuntime, message: Memory, state: State) => {
    const character = runtime.character;

    // Character name
    const agentName = character.name ?? "";

    const room = state.data.room ?? (await runtime.getRoom(message.roomId));
    const deterministicSeed = buildDeterministicSeed([
      "provider:character",
      runtime.agentId,
      character.id ?? "character:none",
      room?.worldId ?? "world:none",
      room?.id ?? message.roomId ?? "room:none",
    ]);

    // Handle bio (random selection from array, resolve {{name}} placeholders)
    const rawBioArray = character.bio ?? [];
    const bioArray = resolveNameInArray(rawBioArray, agentName);
    const bioText =
      bioArray.length > 0
        ? deterministicSample(bioArray, 10, deterministicSeed, "bio").join(" ")
        : "";

    const bio = addHeader(`# About ${agentName}`, bioText);

    // System prompt (resolve {{name}} placeholders)
    const system = resolveNamePlaceholder(character.system ?? "", agentName);

    // Resolve {{name}} in topics
    const resolvedTopics = character.topics
      ? resolveNameInArray(character.topics, agentName)
      : [];

    // Select random topic if available
    const topicString =
      resolvedTopics.length > 0
        ? deterministicPickOne(
            resolvedTopics,
            deterministicSeed,
            "selected-topic",
          ) || null
        : null;

    // postCreationTemplate in core prompts.ts
    // Write a post that is {{adjective}} about {{topic}} (without mentioning {{topic}} directly), from the perspective of {{agentName}}. Do not add commentary or acknowledge this request, just write the post.
    // Write a post that is {{Spartan is dirty}} about {{Spartan is currently}}
    const topic = topicString || "";

    // Format topics list
    const topics =
      resolvedTopics.length > 0
        ? `${agentName} is also interested in ${deterministicSample(
            resolvedTopics.filter((t: string) => t !== topicString),
            5,
            deterministicSeed,
            "topic-list",
          )
            .map((t, index, array) => {
              if (index === array.length - 2) {
                return `${t} and `;
              }
              if (index === array.length - 1) {
                return t;
              }
              return `${t}, `;
            })
            .join("")}`
        : "";

    // Resolve {{name}} in adjectives and select random one
    const resolvedAdjectives = character.adjectives
      ? resolveNameInArray(character.adjectives, agentName)
      : [];
    const adjectiveString =
      resolvedAdjectives.length > 0
        ? deterministicPickOne(
            resolvedAdjectives,
            deterministicSeed,
            "selected-adjective",
          ) || ""
        : "";

    const adjective = adjectiveString || "";

    // Format post examples (resolve {{name}} placeholders)
    const postExamplesArray = character.postExamples
      ? resolveNameInArray(character.postExamples, agentName)
      : [];
    const formattedCharacterPostExamples =
      postExamplesArray.length > 0
        ? deterministicSample(
            postExamplesArray,
            50,
            deterministicSeed,
            "post-examples",
          )
            .map((post) => `${post}`)
            .join("\n")
        : "";

    const characterPostExamples =
      formattedCharacterPostExamples &&
      formattedCharacterPostExamples.replaceAll("\n", "").length > 0
        ? addHeader(
            `# Example Posts for ${agentName}`,
            formattedCharacterPostExamples,
          )
        : "";

    // Format message examples (resolve {{name}} placeholders)
    const messageExamplesArray = character.messageExamples ?? [];
    const formattedCharacterMessageExamples =
      messageExamplesArray.length > 0
        ? deterministicSample(
            messageExamplesArray,
            5,
            deterministicSeed,
            "message-examples",
          )
            .map((group, groupIndex) => {
              const exampleNames = Array.from(
                { length: 5 },
                (_unused, nameIndex) =>
                  deterministicHex(
                    deterministicSeed,
                    `message-example:${groupIndex}:name:${nameIndex}`,
                    8,
                  ),
              );

              return group.examples
                .map((message) => {
                  const messageContent = message.content;
                  const actionsText = messageContent?.actions?.join(", ");
                  const rawText = messageContent?.text ?? "";
                  // Resolve {{name}} in example text content
                  const text = resolveNamePlaceholder(rawText, agentName);
                  let messageString = `${resolveNamePlaceholder(message.name ?? "", agentName)}: ${text}${
                    actionsText ? ` (actions: ${actionsText})` : ""
                  }`;
                  exampleNames.forEach((exName, index) => {
                    const placeholder = `{{name${index + 1}}}`;
                    messageString = messageString.replaceAll(
                      placeholder,
                      exName,
                    );
                  });
                  return messageString;
                })
                .join("\n");
            })
            .join("\n\n")
        : "";

    const characterMessageExamples =
      formattedCharacterMessageExamples &&
      formattedCharacterMessageExamples.replaceAll("\n", "").length > 0
        ? addHeader(
            `# Example Conversations for ${agentName}`,
            formattedCharacterMessageExamples,
          )
        : "";

    const roomType = room?.type;
    const isPostFormat =
      roomType === ChannelType.FEED || roomType === ChannelType.THREAD;

    // Style directions (resolve {{name}} placeholders)
    const characterStyle = character.style;
    const characterStyleAll = characterStyle?.all
      ? resolveNameInArray(characterStyle.all, agentName)
      : [];
    const characterStylePost = characterStyle?.post
      ? resolveNameInArray(characterStyle.post, agentName)
      : [];
    const postDirections =
      characterStyleAll.length > 0 || characterStylePost.length > 0
        ? addHeader(
            `# Post Directions for ${agentName}`,
            [...characterStyleAll, ...characterStylePost].join("\n"),
          )
        : "";

    const characterStyleChat = characterStyle?.chat
      ? resolveNameInArray(characterStyle.chat, agentName)
      : [];
    const messageDirections =
      characterStyleAll.length > 0 || characterStyleChat.length > 0
        ? addHeader(
            `# Message Directions for ${agentName}`,
            [...characterStyleAll, ...characterStyleChat].join("\n"),
          )
        : "";

    const directions = isPostFormat ? postDirections : messageDirections;
    const examples = isPostFormat
      ? characterPostExamples
      : characterMessageExamples;

    const values = {
      agentName,
      bio,
      system,
      topic,
      topics,
      adjective,
      messageDirections,
      postDirections,
      directions,
      examples,
      characterPostExamples,
      characterMessageExamples,
    };

    const data = {
      bio,
      adjective,
      topic,
      topics,
      character,
      directions,
      examples,
      system,
    };

    const topicSentence = topicString
      ? `${agentName} is currently interested in ${topicString}`
      : "";
    const adjectiveSentence = adjectiveString
      ? `${agentName} is ${adjectiveString}`
      : "";
    // Combine all text sections
    const text = [
      bio,
      adjectiveSentence,
      topicSentence,
      topics,
      directions,
      examples,
      system,
    ]
      .filter(Boolean)
      .join("\n\n");

    return {
      values,
      data: {
        bio: data.bio,
        adjective: data.adjective,
        topic: data.topic,
        topics: data.topics,
        character: data.character,
        directions: data.directions,
        examples: data.examples,
        system: data.system,
      },
      text,
    };
  },
};
